client/src/Components/Dialogue/index.tsx

```diff
@@ -0,0 +1,229 @@
+
+  /**
+   * show transform button (shows transform button at the end of the dialogue)
+   */
+  transform?: boolean
+
+  /**
+   * include back button for dialogue
+   */
+  backButton?: boolean
+
+  /**
+   * function for closing the dialogue
+   */
+  closeDialogue?: () => void
+}
+
+/**
+ * Primary UI component for user interaction
+ */
+export const Dialogue: React.FC<DialogueProps> = ({
+  characterName = 'Banana Man',
+  transformName = 'Super Banana Man',
+  dialogueText = [
+    'Hello I am banana man',
+    'welcome to my island',
+    'how are you going today?',
+    'bananas are good for potassium and stuff',
+    'Goodbye I was banana man'
+  ],
+  headerColor = 'orange',
+  avatar = bananaMan,
+  transformAvatar = transformMan,
+  transformAnimation,
+  direction = 'right',
+  transform = 'true',
+  backButton = 'true',
+  closeDialogue
+}) => {
+  const bgColour: 'bg-primary' | 'bg-orange' =
+    headerColor === 'primary' ? 'bg-primary' : 'bg-orange'
+
+  const numMessage = dialogueText.length
+  const [characterTransformed, setCharacterTransformed] = useState(false)
+  const [displayImage, setDisplayImage] = useState(avatar)
+  const [displayName, setDisplayName] = useState(characterName)
+
+  const [displayDialogue, setDisplayDialogue] = useState({
+    currentMessage: 0,
+    typing: true,
+    typed: false
+  })
+
+  const handleClickBack = () => {
+    if (displayDialogue.currentMessage > 0) {
+      setDisplayDialogue({
+        currentMessage: displayDialogue.currentMessage - 1,
+        typing: false,
+        typed: true
+      })
+    }
+  }
+
+  const handleClickNext = () => {
+    if (
+      displayDialogue.currentMessage < numMessage ||
+      (transform && displayDialogue.currentMessage === numMessage)
+    ) {
+      setDisplayDialogue({
+        currentMessage: displayDialogue.currentMessage + 1,
+        typing: true,
+        typed: false
+      })
+    }
+  }
+
+  // changes character image, and character name
+  const transformCharacter = () => {
+    transformAnimation
+    setCharacterTransformed(true)
+    setDisplayImage(transformAvatar)
+    setDisplayName(transformName)
+  }
+
+  const onTypingComplete = () => {
+    setDisplayDialogue({
+      typing: false,
+      typed: true,
+      currentMessage: displayDialogue.currentMessage
+    })
+  }
+
+  const Message = ({ message }: { message: string }) => {
+    if (!displayDialogue.typed) {
+      return (
+        <Typewriter string={message} onComplete={onTypingComplete} delay={20} />
+      )
+    }
+    return <>{message}</>
+  }
+
+  return (
+    <div className='w-full grid grid-cols-2 p-4'>
+      {direction === 'right' && (
+        <div className='col-start-1 col-span-2 sm:col-start-2 sm:col-span-1 -mb-20 md:-mb-40 p-4'>
+          <img src={displayImage} alt={characterName} />
+        </div>
+      )}
+      {direction === 'left' && (
+        <div className='col-start-1 col-span-2 sm:col-start-1 sm:col-span-1 -mb-20 md:-mb-40 p-4'>
```

Comment:
This is a very large block of JSX and there are some common patterns where you check a combination of transform, message count, and character transformed. There is probably a better way to extract out some of the logic into individual render functions.